sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memcpy.asciidoc

```diff
@@ -0,0 +1,161 @@
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `async_memcpy` function
+`sycl::ext::oneapi::experimental::async_memcpy` is a free function
+that  asynchronously copies a number of elements specified by
+`num_elements` of data of type `T` from the source pointer `src` to
+destination pointer `dest`. It also takes a barrier object of type
+`syclex::barrier` as an argument that can be used to wait on the
+completion of the memory copy.
+
+Permitted types for `T` are all scalar and vector types.
+
+This extension provides two versions of `async_memcpy`: with and
+without `Group` template parameter and argument. In the case of the
+group variant, `group_async_memcpy` is issued by all the threads in
+the group. This is a _group function_, as defined in Section 4.17.3
+of the SYCL specification. In the case of the work-item variant,
+`async_memcpy` is issued by the current work-item.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <typename T,  access::address_space DestSpace,
+  access::decorated DestIsDecorated, access::address_space SrcSpace,
+  access::decorated SrcIsDecorated, sycl::memory_scope Scope>
+  void async_memcpy(multi_ptr<T, DestSpace, DestIsDecorated> dest,
+  multi_ptr<T, SrcSpace, SrcIsDecorated> src, size_t numElements,
+  syclex::barrier<Scope> bar);
+
+  template <typename Group, typename T,  access::address_space Space,
+  access::decorated IsDecorated, sycl::memory_scope Scope>
+  void group_async_memcpy(Group g, multi_ptr<T, Space, IsDecorated>
+  dest, multi_ptr<T, Space, IsDecorated> src, size_t numElements,
+  syclex::barrier<Scope> bar);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+=== `async_memcpy` Example
+
+[source,c++]
+----
+using wg_barrier = syclex::barrier<sycl::memory_scope::work_group>;
+auto psrc = multi_ptr<T, sycl::access::address_space::global_space>(src);
+auto pdest = multi_ptr<T, sycl::access::address_space::local_space>(dest);
+
+q.parallel_for(..., [=](sycl::nd_item it) {
+
+  // Allocate memory for and construct the barrier
+  auto* bar = sycl::ext::oneapi::group_local_memory<wg_barrier>(it.get_group(), nthreads);
+
+  async_memcpy(pdest, psrc, N, bar);
+  // Use the barrier
+  bar->arrive_and_wait();
+
+}).wait();
+----
+
+=== `group_async_memcpy` Example
+
+[source,c++]
+----
+using wg_barrier = syclex::barrier<sycl::memory_scope::work_group>;
+auto psrc = multi_ptr<T, sycl::access::address_space::global_space>(src);
+auto pdest = multi_ptr<T, sycl::access::address_space::local_space>(dest);
+
+q.parallel_for(..., [=](sycl::nd_item it) {
+
+  // Allocate memory for and construct the barrier
+  auto* bar = sycl::ext::oneapi::group_local_memory<wg_barrier>(it.get_group(), nthreads);
+
+  group_async_memcpy(it.get_group(), pdest, psrc, N, bar);
```

Comment:
Is the idea that all work-items in the group jointly copy from `psrc` to `pdest`?  If so, there should be a description someplace saying that all work-items must pass the same value for `pdest`, `psrc`, and `N`.  I presume all work-items should also pass the same value for `bar`?

Reply:
In real applications you often need an interface whereby each thread is passed a different pointer:
imagine you are dealing with some complicated strided/and or with offset data that you wish to copy. All of these existing interfaces assume each thread is passed the same start address, and then (although as you mention doc doesn't state yet) the implication is that the impl assigns group items different addresses wrt the start address to copy, up to N items.
This is fine, but if you want to support real applications you at least need an additional interface (perhaps by e.g. introducing a single item/thread "group"), where each thread can be passed a specific pointer by the user.

Reply:
The group variants adds convergence constraints. I will add description that same value of pdest, psrc, N, bar must be passed.
If these are different as @JackAKirk is suggesting, I think the non-group variants should be used.